File: kora-app-annotation-processor/src/main/java/ru/tinkoff/kora/kora/app/annotation/processor/extension/ExtensionResult.java
package ru.tinkoff.kora.kora.app.annotation.processor.extension;

import com.palantir.javapoet.CodeBlock;

import javax.lang.model.element.Element;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.type.ExecutableType;
import javax.lang.model.type.TypeMirror;
import java.util.List;
import java.util.Set;
import java.util.function.Function;

public sealed interface ExtensionResult {

    static GeneratedResult fromExecutable(ExecutableElement constructor) {
        return new ExtensionResult.GeneratedResult(constructor, (ExecutableType) constructor.asType());
    }

    static GeneratedResult fromExecutable(ExecutableElement executableElement, ExecutableType executableType) {
        return new ExtensionResult.GeneratedResult(executableElement, executableType);
    }

    record GeneratedResult(ExecutableElement sourceElement, ExecutableType targetType) implements ExtensionResult {}

    record CodeBlockResult(Element source, Function<CodeBlock, CodeBlock> codeBlock, TypeMirror componentType, Set<String> componentTag, List<TypeMirror> dependencyTypes,
                           List<Set<String>> dependencyTags) implements ExtensionResult {
    }
}

File: kora-app-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kora/app/ksp/extension/ExtensionResult.kt
package ru.tinkoff.kora.kora.app.ksp.extension

import com.google.devtools.ksp.symbol.*
import com.squareup.kotlinpoet.CodeBlock

sealed interface ExtensionResult {
    class GeneratedResult(val constructor: KSFunctionDeclaration, val type: KSFunction) : ExtensionResult

    class CodeBlockResult(
        val source: KSDeclaration,
        val codeBlock: (CodeBlock) -> CodeBlock,
        val componentType: KSType,
        val componentTag: String?,
        val dependencyTypes: List<KSType>,
        val dependencyTags: List<String?>
    ) : ExtensionResult {
    }

    companion object {
        fun fromConstructor(constructor: KSFunctionDeclaration, type: KSClassDeclaration): ExtensionResult {
            return GeneratedResult(
                constructor,
                constructor.asMemberOf(type.asType(listOf()))
            )
        }

        fun fromExecutable(constructor: KSFunctionDeclaration, type: KSFunction): ExtensionResult {
            return GeneratedResult(
                constructor,
                type
            )
        }
    }

}
